File: python/open_guardrail/guards/math_expression_validate.py
"""Validate mathematical expressions."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult


class _MathExpressionValidate:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "math-expression-validate"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        math_blocks = re.findall(
            r"\$[^$]+\$|\\\(.*?\\\)|\\\[.*?\\\]", text
        )
        inline = re.findall(
            r"(?<!\w)[\d(][^a-zA-Z\n]*"
            r"[+\-*/^=][^a-zA-Z\n]*[\d)]",
            text,
        )
        expressions = math_blocks + inline

        for expr in expressions:
            depth = 0
            unbalanced = False
            for ch in expr:
                if ch in "([{":
                    depth += 1
                elif ch in ")]}":
                    depth -= 1
                if depth < 0:
                    issues.append(
                        "Unbalanced parentheses in:"
                        f" {expr[:40]}"
                    )
                    unbalanced = True
                    break
            if not unbalanced and depth > 0:
                issues.append(
                    "Unclosed parenthesis in:"
                    f" {expr[:40]}"
                )

            cleaned = expr.replace("**", "")
            cleaned = cleaned.replace("++", "")
            cleaned = cleaned.replace("--", "")
            if re.search(r"[+\-*/^]{2,}", cleaned):
                issues.append(
                    "Invalid operator sequence in:"
                    f" {expr[:40]}"
                )

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="math-expression-validate",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Math expression issues found"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def math_expression_validate(
    *, action: str = "warn"
) -> _MathExpressionValidate:
    return _MathExpressionValidate(action=action)

File: packages/guards/src/math-expression-validate.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface MathExpressionValidateOptions {
  action: 'block' | 'warn';
}

export function mathExpressionValidate(options: MathExpressionValidateOptions): Guard {
  return {
    name: 'math-expression-validate',
    version: '0.1.0',
    description: 'Validates mathematical expressions',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      const mathBlocks = text.match(/\$[^$]+\$|\\\(.*?\\\)|\\\[.*?\\\]/g) ?? [];
      const inlineExpr = text.match(/(?<!\w)[\d(][^a-zA-Z\n]*[+\-*/^=][^a-zA-Z\n]*[\d)]/g) ?? [];
      const expressions = [...mathBlocks, ...inlineExpr];

      for (const expr of expressions) {
        let depth = 0;
        for (const ch of expr) {
          if (ch === '(' || ch === '[' || ch === '{') depth++;
          if (ch === ')' || ch === ']' || ch === '}') depth--;
          if (depth < 0) {
            issues.push(`Unbalanced parentheses in: ${expr.slice(0, 40)}`);
            break;
          }
        }
        if (depth > 0) {
          issues.push(`Unclosed parenthesis in: ${expr.slice(0, 40)}`);
        }

        if (/[+\-*/^]{2,}/.test(expr.replace(/\*\*/g, '').replace(/\+\+/g, '').replace(/--/g, ''))) {
          issues.push(`Invalid operator sequence in: ${expr.slice(0, 40)}`);
        }
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'math-expression-validate',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
